lang/java/reef-utils/src/test/java/org/apache/reef/util/MultiAsyncToSyncTest.java

```diff
@@ -0,0 +1,225 @@
+    LOG.log(Level.INFO, "Releasing caller...");
+    blocker.release(identifier);
+    return input + 1;
+  }
+}
+
+/**
+ * Use the MultiAsyncToSync class to implement a synchronous API
+ * that uses asynchronous processing internally.
+ */
+final class SynchronousApi {
+  private static final Logger LOG = Logger.getLogger(SynchronousApi.class.getName());
+  private final int incrementerSleepTimeMillis;
+  private final MultiAsyncToSync blocker;
+  private final ExecutorService executor;
+  private final ConcurrentLinkedQueue<FutureTask<Integer>> taskQueue = new ConcurrentLinkedQueue<>();
+  private final AtomicLong idCounter = new AtomicLong(0);
+
+  /**
+   * Parameterize the object as to length of processing time and call timeout.
+   * @param incrementerSleepTimeSeconds Length of time the incrementer sleeps before
+   *                                    performing the increment and returning.
+   * @param timeoutPeriodSeconds The length of time before the call will timeout.
+   */
+  SynchronousApi(final int incrementerSleepTimeSeconds, final long timeoutPeriodSeconds) {
+    this.incrementerSleepTimeMillis = 1000 * incrementerSleepTimeSeconds;
+    this.blocker = new MultiAsyncToSync(timeoutPeriodSeconds, SECONDS);
+    this.executor = Executors.newFixedThreadPool(2);
+  }
+
+  /**
+   * Asynchronously increment the input parameter.
+   * @param input An integer object whose value is to be incremented by one.
+   * @return The input parameter incremented by one or zero for a timeout.
+   * @throws InterruptedException Thread was interrupted by another thread.
+   * @throws ExecutionException An exception was thrown an internal processing function.
+   */
+  public int apiCall(final Integer input) throws InterruptedException, ExecutionException {
+    // Create a future to run the asynchronous processing.
+    final long identifier = idCounter.getAndIncrement();
+    final FutureTask<Integer> task =
+        new FutureTask<>(new AsynchronousIncrementer(input, identifier, incrementerSleepTimeMillis, blocker));
+    taskQueue.add(task);
+
+    LOG.log(Level.INFO, "Running the incrementer...");
+    // Start the task and block the caller until it completes.
+    executor.execute(task);
+    if (blocker.block(identifier)) {
+      LOG.log(Level.INFO, "Call timed out...");
+      // Timeout occurred before the asynchronous processing completed.
+      return 0;
+    }
+    int result = -1;
+    LOG.log(Level.INFO, "Call getting task result...");
+    result = task.get();
+    return result;
+  }
+
+  /**
+   * Insure all test tasks have completed.
+   */
+  public void complete() throws ExecutionException {
+    try {
+      for (final FutureTask<Integer> task : taskQueue) {
+        task.get();
+      }
+    } catch (ExecutionException ee) {
+      if (!(ee.getCause() instanceof InvalidBlockedCallerIdentifierException)) {
+        throw ee;
+      } else {
+        LOG.log(Level.INFO, "Caught exception waiting for completion...", ee);
```

Comment:
I added a timeoutExpected parameter to the function so we only eat the exception when it is expected as part of the test.